File: app/src/main/java/com/nfcalarmclock/dayofweek/NacDayOfWeekPreference.java
package com.nfcalarmclock.dayofweek;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.preference.Preference;
import com.nfcalarmclock.util.dialog.NacDialog;
import com.nfcalarmclock.R;
import com.nfcalarmclock.shared.NacSharedDefaults;
import com.nfcalarmclock.shared.NacSharedPreferences;
import com.nfcalarmclock.system.NacCalendar;

/**
 * Preference that displays the day of week dialog.
 */
public class NacDayOfWeekPreference
	extends Preference
	implements Preference.OnPreferenceClickListener,
		NacDialog.OnShowListener,
		NacDialog.OnDismissListener
{

	/**
	 * Value of days.
	 */
	protected int mValue;

	/**
	 * Shared preferences.
	 */
	protected final NacSharedPreferences mSharedPreferences;

	/**
	 */
	public NacDayOfWeekPreference(Context context)
	{
		this(context, null);
	}

	/**
	 */
	public NacDayOfWeekPreference(Context context, AttributeSet attrs)
	{
		this(context, attrs, 0);
	}

	/**
	 */
	public NacDayOfWeekPreference(Context context, AttributeSet attrs, int style)
	{
		super(context, attrs, style);
		setLayoutResource(R.layout.nac_preference);
		setOnPreferenceClickListener(this);

		this.mSharedPreferences = new NacSharedPreferences(context);
	}

	/**
	 * @return The shared preferences.
	 */
	protected NacSharedPreferences getShared()
	{
		return this.mSharedPreferences;
	}

	/**
	 * @return The summary text.
	 */
	@Override
	public CharSequence getSummary()
	{
		return this.getShared().getDaysSummary();
	}

	/**
	 * Save the selected days when the dialog is dismissed.
	 */
	@Override
	public boolean onDismissDialog(NacDialog dialog)
	{
		NacDayOfWeek dow = ((NacDayOfWeekDialog) dialog).getDayOfWeek();
		this.mValue = NacCalendar.Days.daysToValue(dow.getDays());

		setSummary(this.getSummary());
		persistInt(this.mValue);
		return true;
	}

	/**
	 * @return The default value.
	 */
	@Override
	protected Object onGetDefaultValue(TypedArray a, int index)
	{
		return a.getInteger(index, NacSharedDefaults.getDaysValue());
	}

	/**
	 * Display the dialog when the preference is selected.
	 */
	@Override
	public boolean onPreferenceClick(@NonNull Preference preference)
	{
		Context context = getContext();
		NacDayOfWeekDialog dialog = new NacDayOfWeekDialog();

		dialog.build(context);
		dialog.addOnShowListener(this);
		dialog.addOnDismissListener(this);
		dialog.show();
		return true;
	}

	/**
	 * Set the days in the dialog.
	 */
	@Override
	public void onShowDialog(NacDialog dialog, View root)
	{
		NacDayOfWeek dow = ((NacDayOfWeekDialog) dialog).getDayOfWeek();

		dow.setDays(this.mValue);
	}

	/**
	 * Set the initial preference value.
	 */
	@Override
	protected void onSetInitialValue(Object defaultValue)
	{
		if (defaultValue == null)
		{
			this.mValue = getPersistedInt(this.mValue);
		}
		else
		{
			this.mValue = (Integer) defaultValue;
			persistInt(this.mValue);
		}
	}

}

File: app/src/main/java/com/nfcalarmclock/view/dayofweek/NacDayOfWeekPreference.kt
package com.nfcalarmclock.view.dayofweek

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.fragment.app.FragmentManager
import androidx.preference.Preference
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.util.NacCalendar
import com.nfcalarmclock.util.NacCalendar.Day
import java.util.EnumSet

/**
 * Preference that displays the day of week dialog.
 */
class NacDayOfWeekPreference @JvmOverloads constructor(

	/**
	 * Context.
	 */
	context: Context,

	/**
	 * Attribute set.
	 */
	attrs: AttributeSet? = null,

	/**
	 * Default style.
	 */
	style: Int = 0

	// Constructor
) : Preference(context, attrs, style),

	// Interface
	NacDayOfWeekDialog.OnDaysOfWeekSelectedListener
{

	/**
	 * Value of days.
	 */
	private var dayOfWeekValue = 0

	/**
	 * Shared preferences.
	 */
	private val sharedPreferences: NacSharedPreferences = NacSharedPreferences(context)

	/**
	 * Constructor.
	 */
	init
	{
		layoutResource = R.layout.nac_preference
	}

	/**
	 * Get the summary text.
	 *
	 * @return The summary text.
	 */
	override fun getSummary(): CharSequence
	{
		// Get the default days
		val value = sharedPreferences.days

		// Get the day to start the week on
		val start = sharedPreferences.startWeekOn

		// Convert the default days to a string
		val days = Day.valueToDayString(context, value, start)

		// Return the days string, otherwise return the "None" string
		return days.ifEmpty { context.getString(R.string.none) }
	}

	/**
	 * Called when the days of week are selected.
	 */
	override fun onDaysOfWeekSelected(selectedDays: EnumSet<Day>)
	{
		// Set the day of week value
		dayOfWeekValue = Day.daysToValue(selectedDays)

		// Reevaluate the summary
		summary = this.summary

		// Persist the value
		persistInt(dayOfWeekValue)
	}

	/**
	 * Get the default value.
	 *
	 * @return The default value.
	 */
	override fun onGetDefaultValue(a: TypedArray, index: Int): Any
	{
		// Calculate the default value
		val default = Day.daysToValue(Day.WEEKDAY)

		// Get the default value
		return a.getInteger(index, default)
	}

	/**
	 * Set the initial preference value.
	 */
	override fun onSetInitialValue(defaultValue: Any?)
	{
		// Check if the default value is null
		if (defaultValue == null)
		{
			dayOfWeekValue = getPersistedInt(dayOfWeekValue)
		}
		// Convert the default value
		else
		{
			dayOfWeekValue = defaultValue as Int
			persistInt(dayOfWeekValue)
		}
	}

	/**
	 * Show the start week on dialog.
	 */
	fun showDialog(manager: FragmentManager)
	{
		// Create the dialog
		val dialog = NacDayOfWeekDialog()

		// Setup the dialog
		dialog.defaultDayOfWeekValues = dayOfWeekValue
		dialog.onDaysOfWeekSelectedListener = this

		// Show the dialog
		dialog.show(manager, NacDayOfWeekDialog.TAG)
	}

}